aequilibrae/project/network/osm_builder.py

```diff
@@ -0,0 +1,289 @@
+import math
+from typing import List
+import numpy as np
+from .haversine import haversine
+from aequilibrae import logger
+from aequilibrae.parameters import Parameters
+from PyQt5.QtCore import QObject, pyqtSignal
+
+
+class OSMBuilder(QObject):
+    building = pyqtSignal(object)
+
+    def __init__(self, osm_items: List, conn, node_start=10000) -> None:
+        QObject.__init__(self, None)
+        self.osm_items = osm_items
+        self.conn = conn
+        self.curr = self.conn.cursor()
+        self.node_start = node_start
+        self.report = []
+        self.nodes = {}
+        self.links = {}
+        self.insert_qry = """INSERT INTO {} ({}, geometry) VALUES({}, GeomFromText("{}", 4326))"""
+
+    def doWork(self):
+        node_count = self.data_structures()
+        nodes_to_add, node_ids = self.importing_links(node_count)
+        self.import_nodes(nodes_to_add, node_ids)
+
+    def data_structures(self):
+
+        osmi = []
+        logger.info("Consolidating geo elements")
+        self.building.emit(["text", "Consolidating geo elements"])
+        self.building.emit(["maxValue", len(self.osm_items)])
+
+        for i, x in enumerate(self.osm_items):
+            osmi.append(x["elements"])
+            self.building.emit(["Value", i])
+        self.osm_items = sum(osmi, [])
+
+        logger.info("Separating nodes and links")
+        self.building.emit(["text", "Separating nodes and links"])
+        self.building.emit(["maxValue", len(self.osm_items)])
+
+        alinks = []
+        n = []
+        for i, x in enumerate(self.osm_items):
+            if x["type"] == "way":
+                alinks.append(x)
+            elif x["type"] == "node":
+                n.append(x)
+            self.building.emit(["Value", i])
+
+        self.osm_items = None
+        logger.info("Setting data structures for nodes")
+        self.building.emit(["text", "Setting data structures for nodes"])
+        self.building.emit(["maxValue", len(n)])
+
+        for i, node in enumerate(n):
+            nid = node.pop("id")
+            _ = node.pop("type")
+            self.nodes[nid] = node
+            self.building.emit(["Value", i])
+        del n
+
+        logger.info("Setting data structures for links")
+        self.building.emit(["text", "Setting data structures for links"])
+        self.building.emit(["maxValue", len(alinks)])
+
+        all_nodes = []
+        for i, link in enumerate(alinks):
+            osm_id = link.pop("id")
+            _ = link.pop("type")
+            all_nodes.extend(link["nodes"])
+            self.links[osm_id] = link
+            self.building.emit(["Value", i])
+        del alinks
+
+        logger.info("Finalizing data structures")
+        self.building.emit(["text", "Finalizing data structures"])
+
+        node_count = self.unique_count(np.array(all_nodes))
+
+        return node_count
+
+    def importing_links(self, node_count):
```

Comment:
this method feels pretty long ... can it be broken up at all?

Reply:
I break it up in the next version 0.6.1